# Tennis analysis

### Question or problem definition

We want to see if we can create a prediction model that can predict whether a player will win a match or not, based on different statistics

### Acquire training and testing data

import librairies :

In [1]:
# dataframe manipulation
import pandas as pd
import numpy as np
import random as rnd

# visualization
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

Acquire data from csv :

We want to collect a list of matches with the following information :
- match_id
- date
- surface
- best_of (3 sets win => best_of = 5)
- winner_rank
- loser_rank
- winner_rank_points
- loser_rank_points
- winner_elo_rating
- loser_elo_rating
- winner_age
- loser_age
- winner_height
- loser_height

>SELECT match_id, date, surface, best_of, winner_rank, loser_rank, winner_rank_points, loser_rank_points, winner_elo_rating, loser_elo_rating, winner_age, loser_age, winner_height, loser_height FROM match;

We want to add stats before matches to it :

For example if we want to get aces from nadal given a date we have to do :
> SELECT AVG(ps.p_ace), AVG(ps.o_ace) FROM (SELECT p_matches, p_ace, o_matches, o_ace, match_id, date, player_id FROM player_match_stats_v WHERE player_id=4742 AND date <= '2016/01/01') ps;

final statement to get 2018 matches with their stats :

>SELECT pms.match_id, pms.date, pms.surface, pms.level, pms.round, pms.best_of, pms.player_id, pms.opponent_id, pms.player_rank, pms.opponent_rank, pms.player_elo_rating, pms.opponent_elo_rating, pms.player_age, pms.opponent_age, pms.player_height, pms.opponent_height, pms.outcome, pms.p_matches, pms.o_matches, pms.p_sets, pms.o_sets, pms.p_games, pms.o_games, pms.p_tbs, pms.o_tbs, pms.p_ace, pms.p_df, pms.p_sv_pt, pms.p_1st_in, pms.p_1st_won, pms.p_2nd_won, pms.p_sv_gms, pms.p_bp_sv, pms.p_bp_fc, pms.o_ace, pms.o_df, pms.o_sv_pt, pms.o_1st_in, pms.o_1st_won, pms.o_2nd_won, pms.o_sv_gms, pms.o_bp_sv, pms.o_bp_fc, pms.minutes, pms.p_upsets, pms.o_upsets, ps.p_avg_ace, ps.p_avg_df, ps.p_avg_sv_pt, ps.p_avg_1st_in, ps.p_avg_1st_won, ps.p_avg_2nd_won, ps.p_avg_sv_gms, ps.p_avg_bp_sv, ps.p_avg_bp_fc, ps.o_avg_ace, ps.o_avg_df, ps.o_avg_sv_pt, ps.o_avg_1st_in, ps.o_avg_1st_won, ps.o_avg_2nd_won, ps.o_avg_sv_gms, ps.o_avg_bp_sv, ps.o_avg_bp_fc FROM player_match_stats_v pms LEFT JOIN LATERAL (SELECT ROUND(AVG(p_ace), 2) p_avg_ace, ROUND(AVG(p_df), 2) p_avg_df, ROUND(AVG(p_sv_pt), 2) p_avg_sv_pt, ROUND(AVG(p_1st_in), 2) p_avg_1st_in, ROUND(AVG(p_1st_won), 2) p_avg_1st_won, ROUND(AVG(p_2nd_won), 2) p_avg_2nd_won, ROUND(AVG(p_sv_gms), 2) p_avg_sv_gms, ROUND(AVG(p_bp_sv), 2) p_avg_bp_sv, ROUND(AVG(p_bp_fc), 2) p_avg_bp_fc, ROUND(AVG(o_ace), 2) o_avg_ace, ROUND(AVG(o_df), 2) o_avg_df, ROUND(AVG(o_sv_pt), 2) o_avg_sv_pt, ROUND(AVG(o_1st_in), 2) o_avg_1st_in, ROUND(AVG(o_1st_won), 2) o_avg_1st_won, ROUND(AVG(o_2nd_won), 2) o_avg_2nd_won, ROUND(AVG(o_sv_gms), 2) o_avg_sv_gms, ROUND(AVG(o_bp_sv), 2) o_avg_bp_sv, ROUND(AVG(o_bp_fc), 2) o_avg_bp_fc, player_id FROM player_match_stats_v WHERE date <= pms.date GROUP BY player_id) ps ON (pms.player_id=ps.player_id) WHERE pms.date >= '2018-01-01' AND pms.date <='2018-12-31';

In [5]:
tennis = pd.read_csv('2018 matches/matches.csv', delimiter=';')

In [15]:
tennis.shape

(5886, 64)

When we checked in the database we had only 2974 matches in 2018, we found it odd to have almost twice more data

In [8]:
tennis[tennis['match_id'] == 166985]

match_id        date surface level round  best_of  player_id  \
0      166985  06/04/2018       C     D    RR        3       6030   
390    166985  06/04/2018       C     D    RR        3      11769   

     opponent_id  player_rank  opponent_rank  ...  p_avg_bp_fc  o_avg_ace  \
0          11769         87.0           64.0  ...         7.58       6.20   
390         6030         64.0           87.0  ...         6.85       5.12   

     o_avg_df  o_avg_sv_pt  o_avg_1st_in  o_avg_1st_won o_avg_2nd_won  \
0         2.6        76.18         46.70          33.05         14.95   
390       3.7        85.24         51.61          38.42         18.42   

     o_avg_sv_gms  o_avg_bp_sv  o_avg_bp_fc  
0           12.28         3.33         6.03  
390         13.55         3.88         5.94  

[2 rows x 64 columns]

We can see that we have 2 match stats rows for a single match.

In [9]:
tennis['match_id'].nunique()

2943

That's the number of matches in 2018, so we want to drop duplicates.

In [11]:
tennis_non_duplicates = tennis.drop_duplicates(subset='match_id')

In [13]:
tennis_non_duplicates.shape

(2943, 64)

Once duplicates have been dropped, we can see that we have 2943 matches available for our analysis.

Let's save that to a new csv file :

In [24]:
tennis_non_duplicates.to_csv('2018 matches/non_duplicates_tennis_matches.csv', index=False)

In [25]:
tm_stats = pd.read_csv('2018 matches/non_duplicates_tennis_matches.csv')

In [26]:
tm_stats.head()

match_id        date surface level round  best_of  player_id  opponent_id  \
0    166985  06/04/2018       C     D    RR        3       6030        11769   
1    167084  09/04/2018       C     B   R32        3       6364         6196   
2    167085  09/04/2018       C     B   R32        3       5420         4311   
3    167086  09/04/2018       C     B   R32        3       4291         5902   
4    167087  09/04/2018       C     B   R32        3       4994         3893   

   player_rank  opponent_rank  ...  p_avg_bp_fc  o_avg_ace  o_avg_df  \
0         87.0           64.0  ...         7.58       6.20      2.60   
1         26.0           65.0  ...         6.69       7.55      3.74   
2         93.0           62.0  ...         9.05       6.82      3.03   
3         91.0           61.0  ...         7.98       7.96      3.22   
4         55.0           57.0  ...         7.27       5.12      2.70   

   o_avg_sv_pt  o_avg_1st_in  o_avg_1st_won o_avg_2nd_won  o_avg_sv_gms  \
0        76.18         46.70          33.05         14.95         12.28   
1        82.57         49.97          36.63         16.29         12.95   
2        85.92         50.68          36.20         17.91         13.09   
3        80.63         48.48          35.95         16.59         12.79   
4        77.29         48.82          34.54         14.85         12.05   

   o_avg_bp_sv  o_avg_bp_fc  
0         3.33         6.03  
1         4.18         6.77  
2         4.97         7.82  
3         4.20         6.56  
4         3.75         6.16  

[5 rows x 64 columns]

We acquired the data needed for our analysis !

Now we need to split it in train and test datasets :

In [28]:
from sklearn.model_selection import train_test_split

In [29]:
tm_stats_train, tm_stats_test = train_test_split(tm_stats)

In [32]:
print("train shape : ", tm_stats_train.shape)
print("train test : ", tm_stats_test.shape)

train shape :  (2207, 64)
train test :  (736, 64)


Save those to csv :

In [33]:
tm_stats_train.to_csv('2018 matches/tennis_matches_stats_train.csv', index=False)
tm_stats_test.to_csv('2018 matches/tennis_matches_stats_test.csv', index=False)

Declare new dataframes :

In [34]:
tm_stats_train = pd.read_csv('2018 matches/tennis_matches_stats_train.csv')
tm_stats_test = pd.read_csv('2018 matches/tennis_matches_stats_test.csv')

### Wrangle, prepare and cleanse data

#### Analyse by desribing data

>Describe data to see which features are available

In [36]:
print(tm_stats_train.columns.values)

['match_id' 'date' 'surface' 'level' 'round' 'best_of' 'player_id'
 'opponent_id' 'player_rank' 'opponent_rank' 'player_elo_rating'
 'opponent_elo_rating' 'player_age' 'opponent_age' 'player_height'
 'opponent_height' 'outcome' 'p_matches' 'o_matches' 'p_sets' 'o_sets'
 'p_games' 'o_games' 'p_tbs' 'o_tbs' 'p_ace' 'p_df' 'p_sv_pt' 'p_1st_in'
 'p_1st_won' 'p_2nd_won' 'p_sv_gms' 'p_bp_sv' 'p_bp_fc' 'o_ace' 'o_df'
 'o_sv_pt' 'o_1st_in' 'o_1st_won' 'o_2nd_won' 'o_sv_gms' 'o_bp_sv'
 'o_bp_fc' 'minutes' 'p_upsets' 'o_upsets' 'p_avg_ace' 'p_avg_df'
 'p_avg_sv_pt' 'p_avg_1st_in' 'p_avg_1st_won' 'p_avg_2nd_won'
 'p_avg_sv_gms' 'p_avg_bp_sv' 'p_avg_bp_fc' 'o_avg_ace' 'o_avg_df'
 'o_avg_sv_pt' 'o_avg_1st_in' 'o_avg_1st_won' 'o_avg_2nd_won'
 'o_avg_sv_gms' 'o_avg_bp_sv' 'o_avg_bp_fc']


>Classify those to see which feature is categorical and which is numerical

In [38]:
tm_stats_train.head()

match_id        date surface level round  best_of  player_id  opponent_id  \
0    166298  15/01/2018       H     G   R64        5      11176         6407   
1    167486  21/05/2018       C     B    SF        3       4467        11787   
2    166750  26/02/2018       H     A   R32        3      26578         5055   
3    168956  12/11/2018       H     F    SF        3        644         3819   
4    166498  05/02/2018       C     B    QF        3       5070         5577   

   player_rank  opponent_rank  ...  p_avg_bp_fc  o_avg_ace  o_avg_df  \
0         58.0           53.0  ...         6.32       7.78      3.04   
1         75.0          102.0  ...         7.41       5.02      2.64   
2         82.0           88.0  ...         5.46       5.88      2.88   
3          5.0            3.0  ...         5.98       4.76      2.95   
4         21.0           98.0  ...         7.35       5.47      2.97   

   o_avg_sv_pt  o_avg_1st_in  o_avg_1st_won o_avg_2nd_won  o_avg_sv_gms  \
0        78.38         47.40          33.14         14.52         12.07   
1        80.28         48.17          33.10         15.20         12.28   
2        79.29         50.71          38.04         16.21         12.96   
3        81.01         49.23          34.62         15.71         12.77   
4        80.04         48.56          34.13         15.77         12.33   

   o_avg_bp_sv  o_avg_bp_fc  
0         4.40         7.55  
1         4.43         7.74  
2         3.21         4.83  
3         4.14         7.25  
4         4.25         7.10  

[5 rows x 64 columns]

Categorical variables are :
- Nominal :
    - surface
- Ordinal
    - level
    - p_matches, o_matches
    
Numerical variables are :
- Discrete :
    - best_of
    - player_id, opponent_id
    - player_rank, opponent_rank
    - 

- Continuous : 
    - player_elo_rating, opponent_elo_rating
    - player_age, opponent_age
    - player_height, opponent_height

Mixed types :
- round
- date
- outcome